sdk/core/azure-core/src/main/java/com/azure/core/http/rest/SwaggerMethodParser.java

```diff
@@ -273,21 +274,33 @@ public String setPath(Object[] methodArguments) {
      * @param swaggerMethodArguments the arguments that will be used to create the query parameters' values
      * @param urlBuilder The {@link UrlBuilder} where the encoded query parameters will be set.
      */
+    @SuppressWarnings("unchecked")
     public void setEncodedQueryParameters(Object[] swaggerMethodArguments, UrlBuilder urlBuilder) {
         if (swaggerMethodArguments == null) {
             return;
         }
 
-        for (Substitution substitution : querySubstitutions) {
+        for (QuerySubstitution substitution : querySubstitutions) {
             final int parameterIndex = substitution.getMethodParameterIndex();
             if (0 <= parameterIndex && parameterIndex < swaggerMethodArguments.length) {
                 final Object methodArgument = swaggerMethodArguments[substitution.getMethodParameterIndex()];
-                String parameterValue = serialize(serializer, methodArgument);
-                if (parameterValue != null) {
-                    if (substitution.shouldEncode()) {
-                        parameterValue = UrlEscapers.QUERY_ESCAPER.escape(parameterValue);
+                List<Object> methodArguments = new ArrayList<Object>();
+
+                if (substitution.shouldMergeQueryParams() && methodArgument instanceof List) {
+                    methodArguments.addAll((List<Object>) methodArgument);
+                } else {
+                    methodArguments.add(methodArgument);
+                }
+
+                for (Object methodArgumentUnconverted : methodArguments) {
+                    String parameterValue = serialize(serializer, methodArgumentUnconverted);
+
+                    if (parameterValue != null) {
+                        if (substitution.shouldEncode()) {
+                            parameterValue = UrlEscapers.QUERY_ESCAPER.escape(parameterValue);
+                        }
+                        urlBuilder.addQueryParameter(substitution.getUrlParameterName(), parameterValue);
```

Comment:
One comment I made here appears to have been lost, which was that I would probably not create the intermediate `methodArguments` list, only to then iterate over it one or more times - I would just have appropriate logic in the if/else code blocks do work appropriately in a list or a single-item case.

Reply:
Hi @JonathanGiles, I just forgot to comment on the comment, definitely not lost.

Here we have a situation in which the code from 295-304 is common for both situations, so 289-293 is setup code only. As a result, moving the logic instead to inside if/else will result in duplicated code, which can be done away with with a separate method. 

If the main concern is the list creation for when the methodArgument is not a list, then that would solve the problem - in my experience, that's the case the majority of times. Hence, the solution would be something like:

`
if (substitution.shouldMergeQueryParams() && methodArgument instanceof List) {
    List<Object> list = (List<Object>) methodArgument);
    for(Object object : list) {
        dothework(object);
    }
else {
    dothework(methodArgument);       
}
`

As we can see there's a bit of duplication - calls to `dothework()` in both branches, but at least we avoid creating a list for 1 element.

Please let me know if this is the solution you're looking for, then adding it is trivial. The

Reply:
Yes, that was my thinking. Thanks.